File: python/server/services/auth_service.py
import datetime
import time

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..models.user import User
from ..utils import generate_friend_code, generate_login_code, create_access_token, create_refresh_token, decode_token
from .spotify_oauth import exchange_code, get_spotify_user_info


async def initiate_login(session: AsyncSession):
    login_code = generate_login_code()
    state = generate_login_code()
    expiry = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)
    dummy = User(
        friend_code=generate_friend_code(),
        spotify_state=state,
        login_code=login_code,
        login_code_expiry=expiry,
    )
    session.add(dummy)
    await session.commit()
    return login_code, expiry.timestamp(), state, dummy.id


async def handle_spotify_callback(session: AsyncSession, code: str, state: str):
    result = await session.execute(select(User).where(User.spotify_state == state))
    user = result.scalar_one_or_none()
    if not user:
        return None
    tokens = await exchange_code(code)
    if not tokens:
        return None
    spotify_info = await get_spotify_user_info(tokens["access_token"])
    if not spotify_info:
        return None
    user.spotify_access_token = tokens["access_token"]
    user.spotify_refresh_token = tokens.get("refresh_token", user.spotify_refresh_token)
    user.spotify_token_expiry = datetime.datetime.utcnow() + datetime.timedelta(seconds=tokens.get("expires_in", 3600))
    user.display_name = spotify_info.get("display_name")
    user.username = spotify_info.get("id")
    user.avatar_url = spotify_info["images"][0]["url"] if spotify_info.get("images") else None
    await session.commit()
    return user.login_code


async def redeem_login_code(session: AsyncSession, code: str):
    result = await session.execute(select(User).where(User.login_code == code))
    user = result.scalar_one_or_none()
    if not user or not user.login_code_expiry or user.login_code_expiry < datetime.datetime.utcnow():
        return None
    if not user.spotify_access_token:
        return None
    access_token, access_exp = create_access_token(user.id)
    refresh_token, _ = create_refresh_token(user.id)
    user.access_token = access_token
    user.refresh_token = refresh_token
    user.token_expiry = access_exp
    user.login_code = None
    user.login_code_expiry = None
    await session.commit()
    return {
        "access_token": access_token,
        "refresh_token": refresh_token,
        "timeout": access_exp.timestamp(),
    }


async def refresh_access_token(session: AsyncSession, old_access_token: str, old_refresh_token: str):
    payload = decode_token(old_refresh_token)
    if not payload or payload.get("type") != "refresh":
        return None
    user_id = int(payload["sub"])
    result = await session.execute(select(User).where(User.id == user_id))
    user = result.scalar_one_or_none()
    if not user or user.refresh_token != old_refresh_token:
        return None
    access_token, access_exp = create_access_token(user.id)
    refresh_token, refresh_exp = create_refresh_token(user.id)
    user.access_token = access_token
    user.refresh_token = refresh_token
    user.token_expiry = access_exp
    await session.commit()
    return {
        "token": access_token,
        "refresh_token": refresh_token,
        "timeout": access_exp.timestamp(),
    }


async def check_eligible(session: AsyncSession, token: str):
    payload = decode_token(token)
    if not payload or payload.get("type") != "access":
        return False
    user_id = int(payload["sub"])
    result = await session.execute(select(User).where(User.id == user_id))
    user = result.scalar_one_or_none()
    if not user:
        return False
    return user.token_expiry > datetime.datetime.utcnow() if user.token_expiry else False

File: js/apps/server/src/services/authService.ts
import { prisma } from '../utils/database.js';
import {
  createAccessToken,
  createRefreshToken,
  decodeToken,
  extractBearerToken,
  generateFriendCode,
  generateLoginCode
} from '../utils/tokens.js';
import { SpotifyOAuthService } from './spotifyOauth.js';
import { ApiError } from '../utils/http.js';

const LOGIN_CODE_TTL_MS = 5 * 60_000;

export interface LoginInitiation {
  authUrl: string;
  expiryTimestamp: number;
}

export async function initiateLogin(redirectUri: string): Promise<LoginInitiation> {
  const loginCode = generateLoginCode();
  const state = generateLoginCode();
  const expiry = new Date(Date.now() + LOGIN_CODE_TTL_MS);

  await prisma.user.create({
    data: {
      friendCode: generateFriendCode(),
      displayName: 'Pending',
      username: `pending-${Date.now()}-${Math.floor(Math.random() * 1e6)}`,
      spotifyState: state,
      loginCode,
      loginCodeExpiry: expiry
    }
  });

  return {
    authUrl: SpotifyOAuthService.getAuthorizeUrl(state, redirectUri),
    expiryTimestamp: expiry.getTime()
  };
}

export async function handleSpotifyCallback(
  code: string,
  state: string,
  redirectUri: string
): Promise<string | null> {
  const user = await prisma.user.findFirst({ where: { spotifyState: state } });
  if (!user) return null;

  const tokens = await SpotifyOAuthService.exchangeCode(code, redirectUri);
  if (!tokens?.access_token) return null;

  const profile = await SpotifyOAuthService.getUserInfo(tokens.access_token);
  if (!profile) return null;

  await prisma.user.update({
    where: { id: user.id },
    data: {
      spotifyAccessToken: tokens.access_token,
      spotifyRefreshToken: tokens.refresh_token ?? user.spotifyRefreshToken,
      spotifyTokenExpiry: new Date(Date.now() + (tokens.expires_in ?? 3600) * 1000),
      displayName: profile.display_name ?? user.displayName,
      username: profile.id,
      avatarUrl: profile.images?.[0]?.url ?? null
    }
  });

  return user.loginCode;
}

export interface RedeemResult {
  accessToken: string;
  refreshToken: string;
  timeout: number;
}

export async function redeemLoginCode(loginCode: string): Promise<RedeemResult | null> {
  const user = await prisma.user.findFirst({ where: { loginCode } });
  if (!user || !user.loginCodeExpiry || user.loginCodeExpiry.getTime() < Date.now()) {
    return null;
  }
  if (!user.spotifyAccessToken) return null;

  const access = createAccessToken(user.id);
  const refresh = createRefreshToken(user.id);

  await prisma.user.update({
    where: { id: user.id },
    data: {
      accessToken: access.token,
      refreshToken: refresh.token,
      tokenExpiry: access.expiry,
      loginCode: null,
      loginCodeExpiry: null
    }
  });

  return {
    accessToken: access.token,
    refreshToken: refresh.token,
    timeout: access.expiry.getTime()
  };
}

export interface RefreshResult {
  token: string;
  refreshToken: string;
  timeout: number;
}

export async function refreshTokens(
  oldAccessToken: string,
  oldRefreshToken: string
): Promise<RefreshResult | null> {
  const payload = decodeToken(oldRefreshToken);
  if (!payload || payload.type !== 'refresh') return null;

  const user = await prisma.user.findUnique({ where: { id: payload.sub } });
  if (!user || user.refreshToken !== oldRefreshToken) return null;

  const access = createAccessToken(user.id);
  const refresh = createRefreshToken(user.id);

  await prisma.user.update({
    where: { id: user.id },
    data: {
      accessToken: access.token,
      refreshToken: refresh.token,
      tokenExpiry: access.expiry
    }
  });

  return {
    token: access.token,
    refreshToken: refresh.token,
    timeout: access.expiry.getTime()
  };
}

export async function checkEligible(token: string): Promise<boolean> {
  const payload = decodeToken(token);
  if (!payload || payload.type !== 'access') return false;

  const user = await prisma.user.findUnique({ where: { id: payload.sub } });
  if (!user) return false;
  return user.tokenExpiry ? user.tokenExpiry.getTime() > Date.now() : false;
}

export async function getUserIdFromAuthHeader(header: string | undefined): Promise<string> {
  const token = extractBearerToken(header);
  if (!token) throw new ApiError(401, 'Authorization token missing.');
  const payload = decodeToken(token);
  if (!payload || payload.type !== 'access') {
    throw new ApiError(401, 'Invalid or expired session token.');
  }
  return payload.sub;
}
